src/graft_dag.hpp

```diff
@@ -0,0 +1,161 @@
+  // Check if an edge is from the host, otherwise from the graft.
+  bool IsEdgeFromHost(size_t edge_id) const;
+
+  // ** Contains
+
+  // Checks whether the node is in the host DAG or the graft.
+  bool ContainsNode(const Bitset node_subsplit) const;
+  bool ContainsNode(const size_t node_id) const;
+  // Checks whether the node is in the graft only.
+  bool ContainsGraftNode(const Bitset node_subsplit) const;
+  bool ContainsGraftNode(const size_t node_id) const;
+  // Checks whether the edge is in the host DAG or the graft.
+  bool ContainsEdge(const size_t parent_id, const size_t child_id) const;
+  // Checks whether the edge is in the graft only.
+  bool ContainsGraftEdge(const size_t parent_id, const size_t child_id) const;
+
+  // ** Validity Tests
+
+  // Checks if host and graft are in a valid state.
+  // Specifically, checks that:
+  // - Host DAG is in valid state.
+  // - Each node Id corresponds to its array index.
+  // - Each node is valid. That is, either:
+  //    - Node has zero parents and zero children.
+  //    - Node has 1 or more parents, 1 or more right clade children, and 1 or more left
+  //    clade children.
+  // - All GraftDAG node ids are >= HostNodeCount.
+  bool IsValid() const;
+  // Checks if host and graft are in a valid state to add given node pair.
+  // Specifically, check that:
+  // - The nodes are adjacent.
+  // - The nodes do not add/remove any taxa.
+  // - The parent node has at least one parent.
+  // - Including the child node, each clade of the parent node has at least one child.
+  // - Each clade of the child node has at least 1 child.
+  bool IsValidAddGraftNodePair(const Bitset parent_subsplit,
+                               const Bitset child_subsplit) const;
+
+ protected:
+  // ** Modify DAG
+  // These modifications do not ensure a valid, consistent state for DAG.
+
+  // Add node to the graft.
+  void CreateGraftNode(const Bitset &node_subsplit);
+  // Add edge to the graft.
+  void CreateGraftEdge(const size_t parent_id, const size_t child_id);
+  // Add edge to the graft (Overload for when edge relation is known).
+  void CreateGraftEdge(const size_t parent_id, const size_t child_id,
+                       const bool is_left);
+  // Connect main node to all adjacent nodes in vector.  `ignored_node_id_opt` provides
+  // the option for one adjacent node that will not be connected.
+  void ConnectNodeToAdjacentHostNodes(
+      const size_t main_node_id, const SizeVector adjacent_node_ids,
+      const bool is_main_node_parent, const bool is_left,
+      std::optional<size_t> ignored_node_id_opt = std::nullopt);
+  // Connect node in graft to all viable nodes in host.
+  void ConnectNodeToHost(const size_t node_id, const Bitset node_subsplit,
+                         std::optional<size_t> node_to_ignore = std::nullopt);
+
+ protected:
+  // DAG that the graft is proposed to be connected to.
+  SubsplitDAG &host_dag_;
+  // Nodes and edges in the graft.
+  SubsplitDAGStorage storage_;
```

Comment:
I have a slight preference for calling this `graft_storage_` so it's obvious that it only stores the grafts.

Reply:
Changed name.